x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/AbstractPageMappingToIteratorOperator.java

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.operator;
+
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.TransportVersions;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.core.ReleasableIterator;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+/**
+ * Maps a single {@link Page} into zero or more resulting pages.
+ */
+public abstract class AbstractPageMappingToIteratorOperator implements Operator {
```

Comment:
This implements most of the lookup operator. I made it a separate subclass because:
1. It's easier to test that way.
2. We're going to reuse it for stuff like loading columns.
3. We might be able to reuse it for mv_expand. I'm not sure.
4. I think it's easier to read this way - the bits about managing `Iterator<Page>` or merging `Iterator<Block>` into an existing page gets to live here and there's no need to think about where that iterator comes from. I genuinely can't hold the contents of `BlockHash#lookup` and this class in my had at the same time. So this thing just works with any `Iterator`.